ethexe/consensus/src/connect/mod.rs

```diff
@@ -55,40 +91,88 @@ impl ConsensusService for SimpleConnectService {
         Ok(())
     }
 
-    fn receive_synced_block(&mut self, block: H256) -> Result<()> {
-        let Some(block_data) = self.chain_head.as_ref() else {
-            self.output.push_back(ConsensusEvent::Warning(format!(
-                "Received synced block {block}, but no chain-head was received yet",
-            )));
-
-            return Ok(());
-        };
-
-        if block_data.hash != block {
-            self.output.push_back(ConsensusEvent::Warning(format!(
-                "Received synced block {block} is different from the expected block hash {}",
-                block_data.hash
-            )));
-
-            return Ok(());
+    fn receive_synced_block(&mut self, block_hash: H256) -> Result<()> {
+        if let State::WaitingForSyncedBlock { block } = &self.state
+            && block.hash == block_hash
+        {
+            let validators = self.db.validators(block_hash).ok_or(anyhow!(
+                "validators not found for synced block({block_hash})"
+            ))?;
+            let producer = utils::block_producer_for(
+                &validators,
+                block.header.timestamp,
+                self.slot_duration.as_secs(),
+            );
+
+            self.state = State::WaitingForPreparedBlock {
+                block: block.clone(),
+                producer,
+            };
         }
+        Ok(())
+    }
 
-        self.output
-            .push_back(ConsensusEvent::ComputeBlock(block_data.hash));
-
+    fn receive_prepared_block(&mut self, prepared_block_hash: H256) -> Result<()> {
+        if let State::WaitingForPreparedBlock { block, producer } = &self.state
+            && block.hash == prepared_block_hash
+        {
+            if let Some(index) = self.pending_announces.iter().position(|announce| {
+                announce.address() == *producer && announce.data().block_hash == block.hash
+            }) {
+                let (announce, _) = self
+                    .pending_announces
+                    .remove(index)
+                    .expect("Index must be valid")
+                    .into_parts();
+                self.output
+                    .push_back(ConsensusEvent::ComputeAnnounce(announce));
+                self.state = State::WaitingForBlock;
+            } else {
+                self.state = State::WaitingForAnnounce {
+                    block: block.clone(),
+                    producer: *producer,
+                };
+            };
+        }
         Ok(())
     }
 
-    fn receive_computed_block(&mut self, _block_hash: H256) -> Result<()> {
+    fn receive_computed_announce(&mut self, _announce: AnnounceHash) -> Result<()> {
         Ok(())
     }
 
-    fn receive_block_from_producer(&mut self, _block_hash: SignedProducerBlock) -> Result<()> {
+    fn receive_announce(&mut self, announce: SignedAnnounce) -> Result<()> {
+        debug_assert!(
+            self.pending_announces.len() < MAX_PENDING_ANNOUNCES,
```

Comment:
```suggestion
            self.pending_announces.len() <= MAX_PENDING_ANNOUNCES,
```

Otherwise actual maximum is lower